File: C#/198_HouseRobber.cs
/*

    For a house at index N we must choose between:
        - The max of the array that ends in index N - 2 + value at N
        - value at N - 1.
    
    Eg:
    [1, 2, 3, 1] ->  Pick between 3 + 1 or 2. We pick 4.
     ^  ^  ^
    
    [1,  2,  4,  1] -> Pick between 2 + 1 or 4. We pick 4.
         ^   ^   ^
    
    [1,  2,  4,  4] -> End of array.

*/

public class Solution {
    public int Rob(int[] nums) {
        if (nums.Length == 0) return 0;
        if (nums.Length == 1) return nums[0];
        if (nums.Length == 2) return Math.Max(nums[0], nums[1]);
        
        for (int i = 2; i < nums.Length; i++) {
            nums[i] = Math.Max(nums[i - 2] + nums[i], nums[i - 1]);
            nums[i - 1] = Math.Max(nums[i - 2], nums[i - 1]);
        }
        return nums.Last();
    }
}

File: Python/198_HouseRobber.py
"""

    The Plan:
        - Case number of houses is: 0, 1, 2 we simple check and pick the best choice.
        - For 3 houses and forward, we must pick between:
            * All money made until house N-2 and the current house N
            * All money made until house N-1
        - The value picked is the maximum for house N
        - We must set the maximum for house N-1, which is the max between N-2 and N-1

    [1,2,3,1,6]
        > [1,2,4] we picked between 4 (1 + 3) or 2
        > [1,2,4,4] we picked between 3 (2 + 1) or 4
        > [1,2,4,4,5] we picked between 5 (4 + 1) or 4
        
    [2,1,1,2] example updating the max at N-1
        >[2,1,3]
        >[2,2,3,4]

"""

class Solution:
    def rob(self, nums: List[int]) -> int:
        # Base cases
        if len(nums) == 0: return 0
        if len(nums) == 1: return nums[0]
        if len(nums) == 2: return max(nums[0], nums[1])
        
        for i in range(2, len(nums)):
            nums[i] = max(nums[i-2] + nums[i], nums[i-1])
            nums[i-1] = max(nums[i-1], nums[i-2])
        
        return nums[-1]
        